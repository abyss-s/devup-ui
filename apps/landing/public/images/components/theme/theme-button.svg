<svg width="259" height="140" viewBox="0 0 259 140" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M108 79.3333C108.327 79.3333 108.642 79.4532 108.886 79.6702C109.13 79.8873 109.286 80.1863 109.324 80.5106L109.333 80.6666V82C109.333 82.3398 109.203 82.6667 108.97 82.9138C108.736 83.1609 108.417 83.3096 108.078 83.3295C107.739 83.3494 107.405 83.2391 107.144 83.0209C106.884 82.8028 106.716 82.4934 106.676 82.156L106.667 82V80.6666C106.667 80.313 106.807 79.9739 107.057 79.7238C107.307 79.4738 107.646 79.3333 108 79.3333ZM116.417 76.5466L116.543 76.6573L117.476 77.5906C117.715 77.8306 117.854 78.1525 117.864 78.4911C117.875 78.8297 117.756 79.1596 117.532 79.4137C117.308 79.6678 116.995 79.827 116.658 79.8592C116.321 79.8913 115.984 79.7938 115.716 79.5866L115.591 79.476L114.657 78.5426C114.427 78.3129 114.289 78.0071 114.269 77.6826C114.248 77.3582 114.347 77.0374 114.547 76.7807C114.746 76.524 115.033 76.3491 115.352 76.2888C115.671 76.2285 116.002 76.287 116.281 76.4533L116.417 76.5466ZM101.343 76.6573C101.572 76.8869 101.71 77.1924 101.731 77.5164C101.751 77.8404 101.652 78.1608 101.453 78.4173L101.343 78.5426L100.409 79.476C100.169 79.7151 99.8474 79.8539 99.5088 79.8643C99.1702 79.8746 98.8404 79.7557 98.5863 79.5316C98.3322 79.3076 98.1729 78.9952 98.1408 78.658C98.1087 78.3208 98.2061 77.9839 98.4133 77.716L98.524 77.5906L99.4573 76.6573C99.7074 76.4073 100.046 76.2669 100.4 76.2669C100.754 76.2669 101.093 76.4073 101.343 76.6573ZM97.3333 68.6666C97.6732 68.667 98.0001 68.7971 98.2472 69.0304C98.4943 69.2637 98.643 69.5826 98.6629 69.9218C98.6828 70.2611 98.5724 70.5951 98.3543 70.8557C98.1362 71.1163 97.8268 71.2838 97.4893 71.324L97.3333 71.3333H96C95.6602 71.3329 95.3333 71.2028 95.0862 70.9695C94.8391 70.7362 94.6904 70.4174 94.6704 70.0781C94.6505 69.7388 94.7609 69.4048 94.979 69.1442C95.1971 68.8836 95.5065 68.7161 95.844 68.676L96 68.6666H97.3333ZM120 68.6666C120.34 68.667 120.667 68.7971 120.914 69.0304C121.161 69.2637 121.31 69.5826 121.33 69.9218C121.349 70.2611 121.239 70.5951 121.021 70.8557C120.803 71.1163 120.493 71.2838 120.156 71.324L120 71.3333H118.667C118.327 71.3329 118 71.2028 117.753 70.9695C117.506 70.7362 117.357 70.4174 117.337 70.0781C117.317 69.7388 117.428 69.4048 117.646 69.1442C117.864 68.8836 118.173 68.7161 118.511 68.676L118.667 68.6666H120ZM100.284 60.4133L100.409 60.524L101.343 61.4573C101.582 61.6972 101.721 62.0192 101.731 62.3578C101.741 62.6964 101.622 63.0262 101.398 63.2803C101.174 63.5344 100.862 63.6937 100.525 63.7258C100.187 63.7579 99.8507 63.6605 99.5827 63.4533L99.4573 63.3426L98.524 62.4093C98.2948 62.1795 98.1573 61.8741 98.1372 61.5501C98.1171 61.2262 98.2158 60.9061 98.4149 60.6498C98.614 60.3934 98.8997 60.2185 99.2185 60.1578C99.5373 60.0971 99.8673 60.1547 100.147 60.32L100.284 60.4133ZM117.476 60.524C117.706 60.7536 117.843 61.059 117.864 61.3831C117.884 61.7071 117.786 62.0274 117.587 62.284L117.476 62.4093L116.543 63.3426C116.303 63.5818 115.981 63.7206 115.642 63.7309C115.304 63.7413 114.974 63.6223 114.72 63.3983C114.466 63.1742 114.306 62.8619 114.274 62.5247C114.242 62.1874 114.339 61.8506 114.547 61.5826L114.657 61.4573L115.591 60.524C115.841 60.274 116.18 60.1336 116.533 60.1336C116.887 60.1336 117.226 60.274 117.476 60.524ZM108 56.6666C108.327 56.6667 108.642 56.7866 108.886 57.0036C109.13 57.2206 109.286 57.5196 109.324 57.844L109.333 58V59.3333C109.333 59.6731 109.203 60 108.97 60.2471C108.736 60.4942 108.417 60.6429 108.078 60.6629C107.739 60.6828 107.405 60.5724 107.144 60.3543C106.884 60.1362 106.716 59.8268 106.676 59.4893L106.667 59.3333V58C106.667 57.6463 106.807 57.3072 107.057 57.0572C107.307 56.8071 107.646 56.6666 108 56.6666ZM108 63.3333C109.306 63.3332 110.584 63.7169 111.674 64.4366C112.764 65.1564 113.618 66.1805 114.131 67.3818C114.644 68.5831 114.793 69.9086 114.559 71.1937C114.325 72.4788 113.719 73.6669 112.816 74.6104C111.912 75.5539 110.752 76.2113 109.478 76.5008C108.204 76.7903 106.874 76.6993 105.651 76.239C104.429 75.7787 103.368 74.9694 102.602 73.9116C101.836 72.8539 101.397 71.5943 101.34 70.2893L101.333 70L101.34 69.7106C101.415 67.9943 102.149 66.373 103.39 65.1848C104.63 63.9967 106.282 63.3334 108 63.3333Z" fill="#2F2F2F"/>
<path d="M138 70C138 63.3823 143.364 58.0179 149.982 58.0179H152.06L151.02 59.8152C150.326 61.0135 149.982 62.4381 149.982 64.009C149.982 65.2474 150.256 66.4706 150.785 67.5906C151.313 68.7105 152.083 69.6995 153.04 70.4865C153.996 71.2735 155.115 71.8389 156.316 72.1422C157.516 72.4455 158.769 72.4791 159.985 72.2406L162 71.8488L161.349 73.7971C159.762 78.5515 155.274 81.982 149.982 81.982C143.364 81.982 138 76.6176 138 70Z" fill="#2F2F2F"/>
</svg>
