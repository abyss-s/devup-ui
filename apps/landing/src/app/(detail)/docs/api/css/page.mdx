export const metadata = {
  title: 'css',
}

# css

`css` function is a function that returns a string.

## How to use

### CSS Literal

You can use `css` as a tag function to create a css string. Pass in a string of css properties as an argument, ans it will be converted
into a class list.

```jsx
// Before
<any className={css`
    color: red;
`}/>

// After
<any className="d0"/>
```

### CSS Object

You can also use the css function by passing in a css object as an argument.

```jsx
// Before
<any className={css({
    color: "red"
})}/>

// After
<any className="d0" />
```
