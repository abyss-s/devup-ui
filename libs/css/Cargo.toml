[package]
name = "css"
version = "0.1.0"
edition = "2024"

[dependencies]
once_cell = "1.21.3"
phf = { version = "0.12", features = ["macros"] }
serial_test = "3.2.0"
serde = { version = "1.0.219", features = ["derive"] }
regex = "1.11.1"
