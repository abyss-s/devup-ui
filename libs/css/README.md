## Css

Convert css properties to a string.
