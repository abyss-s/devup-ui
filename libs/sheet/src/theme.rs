use serde::{Deserialize, Deserializer, Serialize};
use std::collections::{BTreeMap, HashMap};

#[derive(Default, Serialize, Deserialize, Debug)]
pub struct ColorTheme(pub HashMap<String, String>);

impl ColorTheme {
    pub fn add_color(&mut self, name: &str, value: &str) {
        self.0.insert(name.to_string(), value.to_string());
    }
}

pub fn deserialize_string_from_number<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum StringOrNumber {
        String(String),
        Number(i64),
        Float(f64),
    }

    match StringOrNumber::deserialize(deserializer)? {
        StringOrNumber::String(s) => Ok(Some(s)),
        StringOrNumber::Number(n) => Ok(Some(n.to_string())),
        StringOrNumber::Float(n) => Ok(Some(n.to_string())),
    }
}
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Typography {
    pub font_family: Option<String>,
    pub font_size: Option<String>,

    #[serde(deserialize_with = "deserialize_string_from_number", default)]
    pub font_weight: Option<String>,
    #[serde(deserialize_with = "deserialize_string_from_number", default)]
    pub line_height: Option<String>,
    pub letter_spacing: Option<String>,
}
impl Typography {
    pub fn new(
        font_family: Option<String>,
        font_size: Option<String>,
        font_weight: Option<String>,
        line_height: Option<String>,
        letter_spacing: Option<String>,
    ) -> Self {
        Self {
            font_family,
            font_size,
            font_weight,
            line_height,
            letter_spacing,
        }
    }
}

#[derive(Serialize, Debug)]
pub struct Typographies(Vec<Option<Typography>>);

impl From<Vec<Option<Typography>>> for Typographies {
    fn from(v: Vec<Option<Typography>>) -> Self {
        Self(v)
    }
}
impl<'de> Deserialize<'de> for Typographies {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum ArrayOrSingle {
            Array(Vec<Option<Typography>>),
            Single(Typography),
        }
        match ArrayOrSingle::deserialize(deserializer)? {
            ArrayOrSingle::Array(v) => Ok(Self(v)),
            ArrayOrSingle::Single(v) => Ok(Self(vec![Some(v)])),
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Theme {
    #[serde(default)]
    pub colors: BTreeMap<String, ColorTheme>,
    #[serde(default = "default_break_points")]
    pub break_points: Vec<u16>,
    #[serde(default)]
    pub typography: BTreeMap<String, Typographies>,
}

fn default_break_points() -> Vec<u16> {
    vec![0, 480, 768, 992, 1280]
}

impl Default for Theme {
    fn default() -> Self {
        Self {
            colors: Default::default(),
            break_points: vec![0, 480, 768, 992, 1280],
            typography: BTreeMap::new(),
        }
    }
}

impl Theme {
    pub fn update_break_points(&mut self, break_points: Vec<u16>) {
        for (idx, value) in break_points.iter().enumerate() {
            let prev = self.break_points.get_mut(idx);
            if let Some(prev) = prev {
                *prev = *value;
            } else {
                self.break_points.push(*value);
            }
        }
    }

    pub fn add_color_theme(&mut self, name: &str, theme: ColorTheme) {
        self.colors.insert(name.to_string(), theme);
    }

    pub fn add_typography(&mut self, name: &str, typography: Vec<Option<Typography>>) {
        self.typography.insert(name.to_string(), typography.into());
    }

    pub fn to_css(&self) -> String {
        let mut theme_declaration = String::new();

        let default_theme_key = self.colors.keys().find(|k| *k == "default").or_else(|| {
            self.colors
                .keys()
                .find(|k| *k == "light")
                .or_else(|| self.colors.keys().next())
        });
        if let Some(default_theme_key) = default_theme_key {
            let entries = {
                let mut col: Vec<_> = self.colors.iter().collect();
                col.sort_by(|a, b| {
                    if a.0 == default_theme_key {
                        std::cmp::Ordering::Less
                    } else if b.0 == default_theme_key {
                        std::cmp::Ordering::Greater
                    } else {
                        a.0.cmp(b.0)
                    }
                });
                col
            };
            let single_theme = entries.len() <= 1;
            for (theme_name, theme_properties) in entries {
                if let Some(theme_key) = if *theme_name == *default_theme_key {
                    None
                } else {
                    Some(theme_name)
                } {
                    theme_declaration.push_str(
                        format!(":root[data-theme={}]{{{}", theme_key, "color-scheme:dark;")
                            .as_str(),
                    );
                } else {
                    theme_declaration.push_str(
                        format!(
                            ":root{{{}",
                            if single_theme {
                                ""
                            } else {
                                "color-scheme:light;"
                            }
                        )
                        .as_str(),
                    );
                }
                for (prop, value) in theme_properties.0.iter() {
                    theme_declaration.push_str(format!("--{}:{};", prop, value).as_str());
                }
                theme_declaration.push_str("}\n");
            }
        }
        let mut css = theme_declaration;
        let mut level_map = BTreeMap::<u8, Vec<String>>::new();
        for ty in self.typography.iter() {
            for (idx, t) in ty.1 .0.iter().enumerate() {
                if let Some(t) = t {
                    let css_content = format!(
                        "{}{}{}{}{}",
                        t.font_family
                            .clone()
                            .map(|v| format!("font-family:{};", v))
                            .unwrap_or("".to_string()),
                        t.font_size
                            .clone()
                            .map(|v| format!("font-size:{};", v))
                            .unwrap_or("".to_string()),
                        t.font_weight
                            .clone()
                            .map(|v| format!("font-weight:{};", v))
                            .unwrap_or("".to_string()),
                        t.line_height
                            .clone()
                            .map(|v| format!("line-height:{};", v))
                            .unwrap_or("".to_string()),
                        t.letter_spacing
                            .clone()
                            .map(|v| format!("letter-spacing:{}", v))
                            .unwrap_or("".to_string())
                    );
                    if css_content.is_empty() {
                        continue;
                    }
                    let typo_css = format!(".typo-{}{{{}}}", ty.0, css_content);
                    level_map
                        .get_mut(&(idx as u8))
                        .map(|v| v.push(typo_css.clone()))
                        .unwrap_or_else(|| {
                            level_map.insert(idx as u8, vec![typo_css]);
                        });
                }
            }
        }
        for (level, css_vec) in level_map {
            if level == 0 {
                css.push_str(css_vec.join("").as_str());
            } else if let Some(media) = self
                .break_points
                .get(level as usize)
                .map(|v| format!("(min-width:{}px)", v))
            {
                css.push_str(format!("\n@media {}{{{}}}", media, css_vec.join("")).as_str());
            }
        }
        css
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn to_css_from_theme() {
        let mut theme = Theme::default();
        let mut color_theme = ColorTheme::default();
        color_theme.add_color("primary", "#000");

        assert_eq!(color_theme.0.keys().count(), 1);

        theme.add_color_theme("default", color_theme);
        let mut color_theme = ColorTheme::default();
        color_theme.add_color("primary", "#fff");
        theme.add_color_theme("dark", color_theme);
        theme.add_typography(
            "default",
            vec![
                Some(Typography::new(
                    Some("Arial".to_string()),
                    Some("16px".to_string()),
                    Some("400".to_string()),
                    Some("1.5".to_string()),
                    Some("0.5".to_string()),
                )),
                Some(Typography::new(
                    Some("Arial".to_string()),
                    Some("24px".to_string()),
                    Some("400".to_string()),
                    Some("1.5".to_string()),
                    Some("0.5".to_string()),
                )),
            ],
        );

        theme.add_typography(
            "default1",
            vec![
                None,
                Some(Typography::new(
                    Some("Arial".to_string()),
                    Some("24px".to_string()),
                    Some("400".to_string()),
                    Some("1.5".to_string()),
                    Some("0.5".to_string()),
                )),
            ],
        );
        let css = theme.to_css();
        assert_eq!(
            css,
            ":root{color-scheme:light;--primary:#000;}\n:root[data-theme=dark]{color-scheme:dark;--primary:#fff;}\n.typo-default{font-family:Arial;font-size:16px;font-weight:400;line-height:1.5;letter-spacing:0.5}\n@media (min-width:480px){.typo-default{font-family:Arial;font-size:24px;font-weight:400;line-height:1.5;letter-spacing:0.5}.typo-default1{font-family:Arial;font-size:24px;font-weight:400;line-height:1.5;letter-spacing:0.5}}"
        );

        assert_eq!(Theme::default().to_css(), "");
        let mut theme = Theme::default();
        theme.add_typography(
            "default",
            vec![Some(Typography::new(None, None, None, None, None))],
        );
        assert_eq!(theme.to_css(), "");

        let mut theme = Theme::default();
        theme.add_color_theme(
            "default",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "dark",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );
        assert_debug_snapshot!(theme.to_css());

        let mut theme = Theme::default();
        theme.add_color_theme(
            "light",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "dark",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );
        assert_debug_snapshot!(theme.to_css());

        let mut theme = Theme::default();
        theme.add_color_theme(
            "a",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "b",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );
        assert_debug_snapshot!(theme.to_css());

        let mut theme = Theme::default();
        theme.add_color_theme(
            "light",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "b",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "a",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );

        theme.add_color_theme(
            "c",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );
        assert_debug_snapshot!(theme.to_css());

        let mut theme = Theme::default();
        theme.add_color_theme(
            "light",
            ColorTheme({
                let mut map = HashMap::new();
                map.insert("primary".to_string(), "#000".to_string());
                map
            }),
        );
        assert_debug_snapshot!(theme.to_css());
    }

    #[test]
    fn update_break_points() {
        let mut theme = Theme::default();
        theme.update_break_points(vec![0, 480, 768, 992, 1280]);
        assert_eq!(theme.break_points, vec![0, 480, 768, 992, 1280]);
        theme.update_break_points(vec![0, 480, 768, 992, 1280, 1600]);
        assert_eq!(theme.break_points, vec![0, 480, 768, 992, 1280, 1600]);
        theme.update_break_points(vec![0, 480, 768, 992, 1280, 1600, 1920]);
        assert_eq!(theme.break_points, vec![0, 480, 768, 992, 1280, 1600, 1920]);
    }
}
