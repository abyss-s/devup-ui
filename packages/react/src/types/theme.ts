/* eslint-disable @typescript-eslint/no-empty-object-type */
export interface DevupThemeColors {}

/* eslint-disable @typescript-eslint/no-empty-object-type */
export interface DevupTheme {}
