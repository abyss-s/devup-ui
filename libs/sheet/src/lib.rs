pub mod theme;

use crate::theme::Theme;
use css::{convert_property, merge_selector, PropertyType, StyleSelector};
use std::cmp::Ordering::{Greater, Less};
use std::collections::{BTreeMap, HashSet};

trait ExtractStyle {
    fn extract(&self) -> String;
}

#[derive(Debug, Hash, Eq, PartialEq)]
pub struct StyleSheetProperty {
    pub class_name: String,
    pub property: String,
    pub value: String,
    pub selector: Option<StyleSelector>,
    pub basic: bool,
}

impl ExtractStyle for StyleSheetProperty {
    fn extract(&self) -> String {
        match convert_property(self.property.as_str()) {
            PropertyType::Single(prop) => {
                format!(
                    "{}{{{}:{}}}",
                    merge_selector(&self.class_name, self.selector.as_ref()),
                    prop,
                    convert_theme_variable_value(&self.value)
                )
            }
            PropertyType::Multi(multi) => format!(
                "{}{{{}}}",
                merge_selector(&self.class_name, self.selector.as_ref()),
                multi
                    .into_iter()
                    .map(|prop| format!("{}:{};", prop, convert_theme_variable_value(&self.value)))
                    .collect::<Vec<String>>()
                    .join("")
            ),
        }
    }
}

fn convert_theme_variable_value(value: &String) -> String {
    if let Some(value) = value.strip_prefix("$") {
        format!("var(--{})", value)
    } else {
        value.to_string()
    }
}

#[derive(Debug, Hash, Eq, PartialEq)]
pub struct StyleSheetCss {
    pub class_name: String,
    pub css: String,
}

impl ExtractStyle for StyleSheetCss {
    fn extract(&self) -> String {
        format!(".{}{{{}}}", self.class_name, self.css)
    }
}

#[derive(Default)]
pub struct StyleSheet {
    /// level -> properties
    pub properties: BTreeMap<u8, HashSet<StyleSheetProperty>>,
    pub css: HashSet<StyleSheetCss>,
    pub theme: Theme,
    theme_declaration: String,
}

impl StyleSheet {
    pub fn add_property(
        &mut self,
        class_name: &str,
        property: &str,
        level: u8,
        value: &str,
        selector: Option<&StyleSelector>,
        basic: bool,
    ) -> bool {
        let prop = StyleSheetProperty {
            class_name: class_name.to_string(),
            property: property.to_string(),
            value: value.to_string(),
            selector: selector.cloned(),
            basic,
        };
        self.properties.entry(level).or_default().insert(prop)
    }

    pub fn add_css(&mut self, class_name: &str, css: &str) -> bool {
        let prop = StyleSheetCss {
            class_name: class_name.to_string(),
            css: css.to_string(),
        };
        self.css.insert(prop)
    }

    pub fn set_theme(&mut self, theme: Theme) {
        let mut theme_declaration = String::new();
        theme_declaration.push_str(theme.to_css().as_str());
        self.theme = theme;
        self.theme_declaration = theme_declaration;
    }

    pub fn create_css(&self) -> String {
        let mut css = self.theme_declaration.clone();
        for (level, props) in self.properties.iter() {
            let mut sorted_props = props.iter().collect::<Vec<_>>();
            sorted_props.sort_by(|a, b| {
                if a.basic == b.basic {
                    match (a.selector.is_some(), b.selector.is_some()) {
                        (true, false) => Greater,
                        (false, true) => Less,
                        (true, true) => {
                            if a.selector == b.selector {
                                if a.property == b.property {
                                    a.value.cmp(&b.value)
                                } else {
                                    a.property.cmp(&b.property)
                                }
                            } else {
                                a.selector.cmp(&b.selector)
                            }
                        }
                        (false, false) => {
                            if a.property == b.property {
                                a.value.cmp(&b.value)
                            } else {
                                a.property.cmp(&b.property)
                            }
                        }
                    }
                } else {
                    b.basic.cmp(&a.basic)
                }
            });

            let inner_css = sorted_props
                .into_iter()
                .map(ExtractStyle::extract)
                .collect::<Vec<String>>()
                .join("");
            if *level == 0 {
                css.push_str(inner_css.as_str());
            } else {
                css.push_str(
                    format!(
                        "\n@media (min-width:{}px){{{}}}",
                        self.theme
                            .break_points
                            .iter()
                            .enumerate()
                            .find(|(idx, _)| (*idx as u8) == *level)
                            .map(|(_, bp)| *bp)
                            .unwrap_or_else(|| self
                                .theme
                                .break_points
                                .last()
                                .cloned()
                                .unwrap_or(0)),
                        inner_css
                    )
                    .as_str(),
                );
            }
        }
        for prop in self.css.iter() {
            css.push_str(&prop.extract());
        }
        css
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_convert_theme_variable_value() {
        assert_eq!(convert_theme_variable_value(&"1px".to_string()), "1px");
        assert_eq!(
            convert_theme_variable_value(&"$var".to_string()),
            "var(--var)"
        );
    }

    #[test]
    fn test_create_css_sort_test() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "background-color", 1, "red", None, false);
        sheet.add_property("test", "background", 1, "some", None, false);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "border", 0, "1px solid", None, false);
        sheet.add_property("test", "border-color", 0, "red", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }
    #[test]
    fn test_create_css_with_selector_sort_test() {
        let mut sheet = StyleSheet::default();
        sheet.add_property(
            "test",
            "background-color",
            1,
            "red",
            Some(&StyleSelector::Postfix("hover".to_string())),
            false,
        );
        sheet.add_property("test", "background-color", 1, "some", None, false);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "background-color", 1, "red", None, false);
        sheet.add_property(
            "test",
            "background-color",
            1,
            "some",
            Some(&StyleSelector::Postfix("hover".to_string())),
            false,
        );
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "background-color", 1, "red", None, false);
        sheet.add_property("test", "background", 1, "some", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }
    #[test]
    fn test_create_css_with_basic_sort_test() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "background-color", 1, "red", None, true);
        sheet.add_property("test", "background", 1, "some", None, false);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "border", 0, "1px solid", None, false);
        sheet.add_property("test", "border-color", 0, "red", None, true);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "display", 0, "flex", None, true);
        sheet.add_property("test", "display", 0, "block", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }

    #[test]
    fn test_create_css_with_selector_and_basic_sort_test() {
        let mut sheet = StyleSheet::default();
        sheet.add_property(
            "test",
            "background-color",
            1,
            "red",
            Some(&StyleSelector::Postfix("hover".to_string())),
            false,
        );
        sheet.add_property("test", "background-color", 1, "some", None, true);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "display", 0, "flex", None, true);
        sheet.add_property("test", "display", 0, "none", None, false);
        sheet.add_property("test", "display", 2, "flex", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }

    #[test]
    fn test_create_css() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 1, "40px", None, false);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_css("test", "display:flex;");
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 2, "40px", None, false);
        sheet.add_property("test", "my", 2, "40px", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }

    #[test]
    fn wrong_breakpoint() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 10, "40px", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }

    #[test]
    fn test_selector_with_prefix() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 1, "40px", Some(&"groupHover".into()), false);
        sheet.add_property("test", "mx", 2, "50px", Some(&"groupHover".into()), false);
        assert_debug_snapshot!(sheet.create_css());
    }

    #[test]
    fn test_theme_selector() {
        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 0, "40px", Some(&"themeDark".into()), false);
        sheet.add_property("test", "my", 0, "40px", Some(&"themeDark".into()), false);
        sheet.add_property("test", "mx", 0, "50px", Some(&"themeLight".into()), false);
        assert_debug_snapshot!(sheet.create_css());

        let mut sheet = StyleSheet::default();
        sheet.add_property("test", "mx", 0, "50px", Some(&"themeLight".into()), false);
        sheet.add_property("test", "mx", 0, "41px", None, false);
        sheet.add_property("test", "mx", 0, "51px", Some(&"themeLight".into()), false);
        sheet.add_property("test", "mx", 0, "42px", None, false);
        assert_debug_snapshot!(sheet.create_css());
    }
}
