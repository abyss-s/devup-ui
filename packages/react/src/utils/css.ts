import { DevupCommonProps } from '../types/props'
import type { DevupSelectorProps } from '../types/props/selector'

export function css(props: DevupCommonProps & DevupSelectorProps): string
export function css(strings: TemplateStringsArray): string
export function css(): string

export function css(
  strings?: TemplateStringsArray | (DevupCommonProps & DevupSelectorProps),
): string {
  if (typeof strings === 'undefined') return ''
  if (Array.isArray(strings)) {
    return strings.join('')
  }
  return strings as string
}
