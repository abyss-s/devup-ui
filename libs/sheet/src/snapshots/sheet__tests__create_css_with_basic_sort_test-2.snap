---
source: libs/sheet/src/lib.rs
expression: sheet.create_css(None)
---
"@layer b;@layer b{.test{border-color:red}}.test{border:1px solid}"
